File: 0060-permutation-sequence/0060-permutation-sequence.java
import java.util.*;

class Solution {
    public String getPermutation(int n, int k) {
        List<List<String>> ans = new ArrayList<>();
        String[] nums = new String[n+1];
        for(int i =1; i <= n;i++)
            nums[i] = Integer.toString(i);
        boolean[] visited = new boolean[n+1];
        backtrack(new ArrayList<>(), n, k, nums, visited, ans);

        List<String> target = ans.get(k-1);
        
        StringBuilder sb =  new StringBuilder();
        for(String str: target) {
            sb.append(str);
        }
        return sb.toString();

    }
    public static void backtrack(List<String> curr, int n , int k , String[] nums, boolean[] visited, List<List<String>> ans) {
        if(curr.size() == n) {
            ans.add(new ArrayList<>(curr));
            return ;
        }
        if(ans.size() == k){
            return;
        }
        for(int i =1; i <= n; i++) {
            if(!visited[i]) {
                visited[i] = true;
                curr.add(Integer.toString(i));
                backtrack(curr, n, k, nums, visited, ans);
                curr.remove(curr.size()-1);
                visited[i] = false;
            }
        }

    }
}

File: 0060-permutation-sequence/0060-permutation-sequence.cpp

#include <iostream>
#include <vector>
#include <string>

using namespace std;


class Solution {
public:
    string getPermutation(int n, int k) {
         vector<int> nums;
        for (int i = 1; i <= n; ++i) {
            nums.push_back(i);
        }
        
        string ans;
        
        while (n > 1) { 
            int page = factorial(n - 1); // (n-1)! permutations per block
            int idx = (k - 1) / page; // find the index in the current list
            k = (k - 1) % page + 1; // adjust k for the next block
            ans.push_back(nums[idx] + '0'); // append the number to the result
            nums.erase(nums.begin() + idx); // remove the used number
            --n;
        }
        ans.push_back(nums[0] + '0'); // append the last remaining number
        return ans;
    }


private:
    int factorial(int n) {
        if (n == 0 || n == 1) return 1;
        int result = 1;
        for (int i = 2; i <= n; ++i) {
            result *= i;
        }
        return result;
    }
};



